### PR TITLE
feat(identity): allow importing and exporting ECDSA keys

### PR DESCRIPTION
**Pull request by drHuangMHT**
## Description

Implement encoding to/decoding from DER-encoded secret key document for `ecdsa::SecretKey`.
Implement encoding to/decoding from protobuf format for ECDSA keys.
Bump dependency `p256` from 0.12 to 0.13.  
Bump dependency `sec1` from 0.3.0 to 0.7

Related: #3681.

## Notes & open questions

~~Attempt for decoding support failed because the returned error type `pkcs8::error::Error` does not implement `std::error::Error`(requires `std` feature on `pkcs8`, despite `std` is enabled on `p256`), which is required for `DecodingError::failed_to_parse`.~~

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
> ``(requires `std` feature on `pkcs8`, despite `std` is enabled on `p256`)

You can work around this by adding an explicit dependency for `pkcs8` and activate the `std` feature there. I'd also recommend opening an upstream issue/PR that fixes this so we can eventually remove that dependency again!

**Review by thomaseizinger — COMMENTED** [review]
Awesome! Thank you for splitting this out.

I think we should integrate this into the `Keypair::to_protobuf_encoding` function.
Also, can you add a changelog entry?

**Comment by drHuangMHT** [comment]
> > ``(requires `std` feature on `pkcs8`, despite `std` is enabled on `p256`)
> 
> You can work around this by adding an explicit dependency for `pkcs8` and activate the `std` feature there. I'd also recommend opening an upstream issue/PR that fixes this so we can eventually remove that dependency again!

Can be fixed by using `p256` version 0.13

**Comment by drHuangMHT** [comment]
By the way, we can safely drop dependency `sec1` now.

**Review by thomaseizinger — COMMENTED** [review]
Solid, thank you for this!

Can you update `Keypair`'s protobuf encoding functions to use these new methods?

**Comment by thomaseizinger** [comment]
> By the way, we can safely drop dependency `sec1` now.

Great news! Want to send a PR?

**Comment by drHuangMHT** [comment]
> Solid, thank you for this!
> 
> Can you update `Keypair`'s protobuf encoding functions to use these new methods?

I think we can do that in a separate PR, but if you say so

**Comment by thomaseizinger** [comment]
> > Solid, thank you for this!
> > Can you update `Keypair`'s protobuf encoding functions to use these new methods?
> 
> I think we can do that in a separate PR, but if you say so

We could but it seems unnecessary fine-granular. These methods aren't used outside of the protobuf encoding. Thinking about it, we should probably make them `pub(crate)`.

**Comment by drHuangMHT** [comment]
> > By the way, we can safely drop dependency `sec1` now.
> 
> Great news! Want to send a PR?

I'm not sure how to get rid of those commits on my `master` branch though. Maybe I should make a new fork of this repo?

**Comment by drHuangMHT** [comment]
> We could but it seems unnecessary fine-granular. These methods aren't used outside of the protobuf encoding. Thinking about it, we should probably make them `pub(crate)`.

What if users want to import keys? Can be achieved by using `From` impl but I see none.

**Comment by thomaseizinger** [comment]
> > > By the way, we can safely drop dependency `sec1` now.
> > 
> > 
> > Great news! Want to send a PR?
> 
> I'm not sure how to get rid of those commits on my `master` branch though. Maybe I should make a new fork of this repo?

You can make a new branch on your fork and send a PR from that one :)

**Comment by thomaseizinger** [comment]
> > We could but it seems unnecessary fine-granular. These methods aren't used outside of the protobuf encoding. Thinking about it, we should probably make them `pub(crate)`.
> 
> What if users want to import keys? Can be achieved by using `From` impl but I see none.

I am bit reluctant to add APIs for such "just in case" usecases. All of https://github.com/libp2p/rust-libp2p/issues/3630 is about important and exporting keys. We do that with a format that is supported across all libp2p implementations. From where other than other libp2p implementations do you want to import a key?

Historically, this repository has had and still has a lot symbols in its public API that are actually unused by us but need to be maintained and limit us from making changes because they could be breaking for our users. I've been working on deprecating/removing them for a while, see https://github.com/libp2p/rust-libp2p/pull/3806 as an example. This was/is a lot of work that I'd rather spend implementing features.

Sorry to be picky here but I unless we have a validated usecase for an API, I'd rather not add it to ease our maintenance load down the line. I hope you understand.

**Comment by drHuangMHT** [comment]
> Sorry to be picky here but I unless we have a validated usecase for an API, I'd rather not add it to ease our maintenance load down the line. I hope you understand.

Not even `From<p256::SigningKey>`?

**Comment by drHuangMHT** [comment]
> > I'm not sure how to get rid of those commits on my `master` branch though. Maybe I should make a new fork of this repo?
> 
> You can make a new branch on your fork and send a PR from that one :)

But the new branch will still contain those commits just like you see in this PR, that's the problem if I simply branch off `master` because they're on `master`.

**Comment by thomaseizinger** [comment]
> > > I'm not sure how to get rid of those commits on my `master` branch though. Maybe I should make a new fork of this repo?
> > 
> > 
> > You can make a new branch on your fork and send a PR from that one :)
> 
> But the new branch will still contain those commits just like you see in this PR, that's the problem if I simply branch off `master` because they're on `master`.

You can branch of `<remote>/master`. Assuming this repository is added as the remote `origin`, the following should do:

```
git checkout -b origin/master remove-sec1-dependency
```

That should give you a branch `remove-sec1-dependency` that is based off _our_ master.

**Comment by thomaseizinger** [comment]
> > Sorry to be picky here but I unless we have a validated usecase for an API, I'd rather not add it to ease our maintenance load down the line. I hope you understand.
> 
> Not even `From<p256::SigningKey>`?

That would add `p256` to our public API, meaning any breaking change to it would be a breaking change for us. `multiaddr` will soon [have a dependency on `libp2p-identity`](https://github.com/multiformats/rust-multiaddr/pull/83). `Multiaddr` on the other hand is a fundamental type of the libp2p ecosystem. A breaking change to it triggers a cascade of breaking changes for pretty much everyone which is very costly.

I am not fundamentally against adding these APIs. I just don't want to do so on a "what if" basis. Can you explain your usecase / application that would require importing keys from external sources?

**Comment by drHuangMHT** [comment]
> > > Sorry to be picky here but I unless we have a validated usecase for an API, I'd rather not add it to ease our maintenance load down the line. I hope you understand.
> > 
> > Not even `From<p256::SigningKey>`?
> 
> That would add `p256` to our public API, meaning any breaking change to it would be a breaking change for us. `multiaddr` will soon [have a dependency on `libp2p-identity`](https://github.com/multiformats/rust-multiaddr/pull/83). `Multiaddr` on the other hand is a fundamental type of the libp2p ecosystem. A breaking change to it triggers a cascade of breaking changes for pretty much everyone which is very costly.

I don't see, or rather can't see any problem behind this unfortunately. Make it crate-level only then.

> I am not fundamentally against adding these APIs. I just don't want to do so on a "what if" basis. Can you explain your usecase / application that would require importing keys from external sources?

No. I'm just making things up.

**Comment by thomaseizinger** [comment]
Okay, glad to have resolved this. If a usecase comes up, feel free to open an issue!

(I also just saw that we have a public `try_from_bytes` already anyway 🙂).

To merge this, could you please:

- Make the new functions `pub(crate)`
- Use them in the Keypair protobuf functions

**Comment by drHuangMHT** [comment]
> (I also just saw that we have a public `try_from_bytes` already anyway 🙂).

I thought that one will also be made visible only on crate-level. I guess there's no other usecase?

**Comment by drHuangMHT** [comment]
> You can branch of `<remote>/master`. Assuming this repository is added as the remote `origin`, the following should do:
> 
> ```
> git checkout -b origin/master remove-sec1-dependency
> ```
> 
> That should give you a branch `remove-sec1-dependency` that is based off _our_ master.

Actually the remote origin is on my account so that won't work unfortunately. I guess the workflow should be  
1. fork the repository you want to contribute to to your own account.
2. clone the repo on your account to your computer.
3. create a new branch and do the changes there.
4. file a pull request from your branch on your own repo to the repo for contributing.  

I forgot to create a new branch previously and contaminated `master` whose upstream is this repo and any update here will result in an additional merge commit that does not exist on this repo. I can simply drop those changes but I'm afraid something bad will happen. Just YOLO it?

**Comment by thomaseizinger** [comment]
> Actually the remote origin is on my account so that won't work unfortunately.

You can add a new one:

```
git remote add upstream git@github.com:libp2p/rust-libp2p.git
```

And then do:

```
git fetch upstream
git checkout -b upstream/master remove-sec1-dependency
```

**Comment by thomaseizinger** [comment]
> I forgot to create a new branch previously

Another way would be to simply branch off a different commit:

```
git checkout -b remove-sec1-dependency <COMMIT_HASH>
```

Where <COMMIT_HASH> is a commit before all your changes. Find one with `git log`.

File: identity/CHANGELOG.md [hunk 1]
```diff
@@ -2,8 +2,11 @@
 
 - Raise MSRV to 1.65.
   See [PR 3715].
+- Add PKCS#8 encoding/decoding support for ECDSA keys.
+  See [PR 3863].
 
 [PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
+[PR 3863]: https://github.com/libp2p/rust-libp2p/pull/3863
```
Comment (thomaseizinger) [inline identity/CHANGELOG.md]:
```suggestion
- Add support for exporting and importing ECDSA keys via the libp2p [protobuf format].
  See [PR 3863].

[PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
[PR 3863]: https://github.com/libp2p/rust-libp2p/pull/3863
[protobuf format]: https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#keys
```

**Review by thomaseizinger — DISMISSED** [review]
Implementation looks good to me, CI failures need to be addressed :)

File: identity/src/ecdsa.rs [hunk 2]
```diff
@@ -285,4 +308,12 @@ mod tests {
         let invalid_msg = "h3ll0 w0rld".as_bytes();
         assert!(!pk.verify(invalid_msg, &sig));
     }
+
+    #[test]
+    fn secret_key_encode_decode_roundtrip() {
```
Comment (thomaseizinger) [inline identity/src/ecdsa.rs]: Can you replace this test with a roundtrip test for an ECDSA key using the `Keypair::to_protobuf_encoding`? Plus one "test vector", i.e. parse hex-encoded protobuf bytes and assert that we can generic a particular `Peerid` from it.

The same way we did it in https://github.com/libp2p/rust-libp2p/pull/3681.

Comment (drHuangMHT) [reply identity/src/ecdsa.rs]: I guess that's what you expect with `1c8d282`?

Comment (thomaseizinger) [reply identity/src/ecdsa.rs]: Yep, thanks! I adapted the test vectors to actually be correct.

**Review by thomaseizinger — DISMISSED** [review]
Awesome, thank you!

Waiting with merge on the otherlibp2p implementations to check that we are doing the right thing:

See https://github.com/libp2p/specs/pull/537.

**Comment by drHuangMHT** [comment]
Um what's wrong with the hex value?

**Comment by thomaseizinger** [comment]
We don't seem to be doing the correct encoding, can you look into that?

https://github.com/libp2p/specs/pull/537#discussion_r1183502385

**Comment by drHuangMHT** [comment]
> We don't seem to be doing the correct encoding, can you look into that?
> 
> [libp2p/specs#537 (comment)](https://github.com/libp2p/specs/pull/537#discussion_r1183502385)

The key used for testing is generated using openssl. It was in pem so I told openssl to convert it to pkcs#8. The first integer is 0 by looking at the key on my machine. However, the key generated by our implementation also comes with the first integer being 0. `p256` is a pure rust implementation right? So could it be `p256`? Still digging.

**Comment by drHuangMHT** [comment]
Problem solved. However I don't see a convenient way to encode the key. Seems we need to implement it ourselves.

**Comment by thomaseizinger** [comment]
> Problem solved. However I don't see a convenient way to encode the key. Seems we need to implement it ourselves.

I haven't dug into the details but you mentioned RFC5915 in the specs repository. A crates.io search revealed the following: https://docs.rs/sec1/0.7.2/sec1/

Perhaps that is all we need?

**Comment by drHuangMHT** [comment]
> > Problem solved. However I don't see a convenient way to encode the key. Seems we need to implement it ourselves.
> 
> I haven't dug into the details but you mentioned RFC5915 in the specs repository. A crates.io search revealed the following: https://docs.rs/sec1/0.7.2/sec1/
> 
> Perhaps that is all we need?

God we just wanted to drop that dependency, hilarious.

**Review by thomaseizinger — APPROVED** [review]
Awesome, thanks for updating!

**Pull request merged**